src/topology/config/component.rs

```diff
@@ -0,0 +1,180 @@
+use inventory;
+use serde::de::{Deserializer, Error, MapAccess, Visitor};
+use serde::ser::Serializer;
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+use std::fmt;
+use toml::Value;
+
+/// Combines a type field and a nested plugin config to create a table.
+#[derive(Serialize, Debug, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct ConfigSwapOut {
+    #[serde(rename = "type")]
+    pub type_str: String,
+    #[serde(flatten)]
+    pub nested: Value,
+}
+
+impl ConfigSwapOut {
+    pub fn new() -> ConfigSwapOut {
+        ConfigSwapOut {
+            type_str: "".to_owned(),
+            nested: Value::Table(BTreeMap::new()),
+        }
+    }
+}
+
+impl<'de> Visitor<'de> for ConfigSwapOut {
+    type Value = ConfigSwapOut;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a map")
+    }
+
+    fn visit_map<M>(self, mut access: M) -> Result<Self::Value, M::Error>
+    where
+        M: MapAccess<'de>,
+    {
+        let mut type_str = "".to_owned();
+        let mut nested: BTreeMap<String, Value> = BTreeMap::new();
+
+        while let Some((key, value)) = access.next_entry::<String, toml::Value>()? {
+            if key == "type" {
+                type_str = value.as_str().unwrap_or("").to_owned();
+            } else {
+                nested.insert(key, value);
+            }
+        }
+
+        if type_str.len() == 0 {
+            Err(Error::custom("missing type field"))
+        } else {
+            Ok(ConfigSwapOut {
+                type_str: type_str,
+                nested: Value::Table(nested),
+            })
+        }
+    }
+}
+
+/// Stores both a constructed plugin instance and the ConfigSwapOut used to
+/// create it, this allows us to echo back the config of a component at runtime.
```

Comment:
I was thinking this would sit under `topology::config::Config` to replace the existing config structs (`TransformOuter` etc) and then the underlying `component` gets pulled out after successful parsing.